Add structured datasets loading capability in valkey benchmark

VoletiRam opened the pull request:
## Background

Add structured datasets loading capability. Support CSV and TSV file formats. Use `__field:fieldname__` placeholders to replace the corresponding fields from the dataset file. Support natural content size of varying length. Allow mixed placeholder usage combining dataset fields with random generators. Enable automatic field discovery from CSV/TSV headers. Use `--maxdocs` to limit the dataset loading.

Rather than modifying the existing placeholder system, we detect field placeholders and switch to a separate code path that builds commands from scratch using `valkeyFormatCommandArgv()`. This ensures:

- Zero impact on existing functionality
- Full support for variable-size content
- Thread-safe atomic record iteration
- Compatible with pipelining and threading modes

__Usage examples__

```sh
# Strings - Simple key-value with dataset fields
./valkey-benchmark --dataset products.csv -n 10000 SET product:__rand_int__ "__field:name__"

# Sets - Unique collections from dataset
./valkey-benchmark --dataset categories.csv -n 10000 SADD tags:__rand_int__ "__field:category__"

# CSV dataset with document limit
./valkey-benchmark --dataset wiki.csv --maxdocs 100000 -n 50000 HSET doc:__rand_int__ title "__field:title__" body "__field:abstract__"

# Mixed placeholders (dataset + random)
./valkey-benchmark --dataset terms.csv -r 5000000 -n 50000 HSET search:__rand_int__ term "__field:term__" score __rand_1st__
```

__Full-Text Search Benchmarking__

```sh
# Search hit scenarios (existing terms)
./valkey-benchmark --dataset search_terms.csv -n 50000 FT.SEARCH rd0 "__field:term__"

# Search miss scenarios (non-existent terms)
./valkey-benchmark --dataset miss_terms.csv -n 50000 FT.SEARCH rd0 "__field:term__"

# Query variations
./valkey-benchmark --dataset search_terms.csv -n 50000 FT.SEARCH rd0 "@title:__field:term__"
./valkey-benchmark --dataset search_terms.csv -n 50000 FT.SEARCH rd0 "__field:term__*"
```

__Benchmark Results__


Test environment:
__Instance:__ AWS c7i.16xlarge, 64 vCPU

Test Dataset: 5M+ Wikipedia XML documents, 5.8GB memory

| Configuration | Throughput | CPU Usage | Wall Time | Memory Peak |
|---------------|------------|-----------|-----------|-------------|
| Single-threaded, P1 | 93,295 RPS | 99% | 71.4s | 5.8GB |
| Multi-threaded (10), P1 | 93,332 RPS | 137% | 71.5s | 5.8GB |
| Single-threaded, P10 | 274,499 RPS | 96% | 36.1s | 5.8GB |
| Multi-threaded (4), P10 | 344,589 RPS | 161% | 32.4s | 5.8GB |

Commit:
Add structured datasets loading capability in valkey benchmark

Add structured datasets loading capability. Support XML/CSV/TSV file
formats. Use `__field:fieldname__' placeholders to replace the
corresponding fields from the dataset file. Support natural content size of
varying length. Allow mixed placeholder usage combining dataset fields with
random generators. Enable automatic field discovery from CSV/TSV headers
and XML tags. Use `--maxdocs` to limit the dataset loading.

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix memory leak in memory reporting

Fix memory leak in memory reporting

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Add documentation for valkey-benchmark and improve xml field discovery

Add documentation for valkey-benchmark. Fix field discovery. Improve
field discovery in xml  and validate field before loading the data.
fix failed tcl test in ubuntu.

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

JimB123 commented on src/valkey-benchmark.c:
```diff
@@ -1928,7 +2604,9 @@ int parseOptions(int argc, char **argv) {
-        "                    the 'fcall' test. (default 1)\n",
+        "                    the 'fcall' test. (default 1)\n"
+        " --dataset <file>   Path to CSV/TSV/XML dataset file for field placeholder replacement.\n"
+        "                    All fields auto-detected with natural content lengths.\n",
```
Have we reached the point where we need some actual documentation for valkey-benchmark?

The benchmark tool has relied on `--help` for increasingly complex configurations.  Now, we're introducing a dataset configuration file with very limited description.  Is it likely that developers will understand how to use this without examining the code?

I suggest that a new documentation file be created (benchmark.md) which can provide details and examples for using valkey-benchmark, including details about this new configuration file.

VoletiRam replied:
Thank you for the suggestion. I agree `--help` is not clear enough for most of the configurations benchmark support. Will add the detailed explanation with examples.

JimB123 reviewed (changes requested):
Need a full documentation file.  Need details on the new file format.  Examples.

Maybe seed the file with --help data for the existing cases.  This can be an area for future improvement.

Commit:
Merge branch 'unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix memory leak in xml scan

Fix memory leak in xml scan

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

JimB123 commented:
Nice.  I appreciate the benchmark.md file.  This is an excellent start and provides a place for future documentation.

JimB123 reviewed (approved):
I really like the new benchmark.md file

rainsupreme commented on docs/benchmark.md:
This doc is great! I noticed `--warmup` and `--duration` options weren't documented. Are there other options not covered? I'd like to either document all of them or have a list of undocumented options here. :)

VoletiRam replied:
Done.

madolson replied:
This file is good, but it doesn't really make sense organizationally. We have documentation here: https://valkey.io/topics/benchmark/ which has it's authoritative source in valkey-doc.

rainsupreme replied:
Oh right, tbh I forgot about that. Organization is good, but we should fix that doc. It's out of date, for both these changes and changes I made myself previously. I guess we should make an issue about it or something? Should the valkey-benchmark help text reference that documentation perhaps? 🤔

madolson replied:
An issue is reasonable. I just don't want two copies of the same information. We can probably move most of this content as is over. A readme in the source tree is not a good place for end user docs.

VoletiRam replied:
I created issue https://github.com/valkey-io/valkey-doc/issues/414 for the update of documentation. I will own it. Please assign it to me.

Commit:
Load fields only we care about from dataset

Load fields only we care about from dataset. Parse the fields from
benchmark command and load only those field values from dataset

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Remove warning log for incomplete document

Remove warning log for incomplete document

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme commented on src/valkey-benchmark.c:
```diff
@@ -1606,6 +1668,816 @@ static void updateClusterSlotsConfiguration(void) {
+    size_t before_len = field_pos - arg;
+    const char *after_start = field_end + FIELD_SUFFIX_LEN;
+
+    sds result = sdsnewlen(arg, before_len);
```
I dunno about the perf here. This is run for every argument in every command sent, and I notice it calls `strstr()` and `sdsnewlen()` which would typically have O(n) runtime. One can argue that this is effectively constant runtime if the user is reasonable about argument length, but it's still a lot of work to be doing in an inner loop. Also, this appears to be redundant work. My ideas:
- The template we're given doesn't change. We could preprocess more and only do the minimum work here for each command sent.
- There is a lot of allocator work going on here - the caller duplicates the template array for every command plus here we replace some number of those allocations again. The benchmark would run faster if we avoided so many allocations. Perhaps we could allocate a large block of memory that is large enough for the longest command we expect and keep reusing that?

If we don't make this more efficient the user will have to be more careful that the benchmark doesn't get too bottlenecked by valkey-benchmark itself, instead of valkey-server. Users have to do this anyway to some extent, but dataset mode demands much more resources than valkey-benchmark usually requires.

VoletiRam replied:
Appreciate the concern on performance. I think about this and definitely gonna be 20-30% improvements in terms of performance with your suggestions. The performance with current change is good enough for full text search needs and saturating the server as desired. I can follow up as future improvements for template array based pre-processing of commands for performance when we know it is a bottleneck.

rainsupreme commented on tests/integration/valkey-benchmark.tcl:
```diff
@@ -199,6 +199,224 @@ tags {"benchmark network external:skip logreqres:skip"} {
+
+        test {benchmark: dataset XML with field placeholders} {
+            # Create test XML dataset matching Wikipedia structure
+            set xml_data "<doc><title>XML Title 1</title><abstract>XML Abstract 1</abstract><url>http://example1.com</url><links><sublink><anchor>test1</anchor><link>http://test1.com</link></sublink></links></doc>\n<doc><title>XML Title 2</title><abstract>XML Abstract 2</abstract><url>http://example2.com</url><links><sublink><anchor>test2</anchor><link>http://test2.com</link></sublink></links></d ...
```
Could we add whitespace formatting to this so it's more readable?

VoletiRam replied:
Done.

rainsupreme reviewed (commented):
Finally finished reviewing - sorry for the delay! I still think having the dataset stuff in a separate file is a good idea. We can write unit tests for all the parsing and replacement stuff too. :)

I don't think the perf concerns are necessarily blockers, but I want to make sure we've thought about it and are reasonably sure it won't be an issue.

Commit:
Separate dataset changes into new file

Separate dataset changes into new file. Add unit test coverage.

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Merge remote-tracking branch 'fork/unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix build

Fix build

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix cmake build issue for latest ubuntu-cmake

Fix cmake build issue for latest ubuntu-cmake

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Replace placeholders with in same field

Replace placeholders with in same field. Update documentation

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme commented on src/unit/CMakeLists.txt:
```diff
@@ -23,7 +23,7 @@ add_library(valkeylib STATIC ${VALKEY_SERVER_SRCS})
 target_compile_definitions(valkeylib PRIVATE "${COMPILE_DEFINITIONS}")
 
-add_executable(valkey-unit-tests ${UNIT_TEST_SRCS})
+add_executable(valkey-unit-tests ${UNIT_TEST_SRCS} ${CMAKE_SOURCE_DIR}/src/valkey-benchmark-dataset.c)
```
I'm not familiar with cmake, but his seems a bit odd and I'm confused. Shouldn't it be part of a list of files like UNIT_TEST_SRCS or something? And why added to valkey-unit-tests but not valkey-benchmark?

rainsupreme commented on src/unit/test_dataset.c:
```diff
@@ -0,0 +1,115 @@
+/* Unit tests for valkey-benchmark dataset module
+ *
+ * Copyright (c) 2024, Redis Ltd.
```
we're not redis

rainsupreme replied:
looks like copyright isn't fixed yet? It's an easy fix but licensing and copyright are important :p

rainsupreme commented on src/Makefile:
```diff
@@ -498,14 +521,14 @@ endif
 
 # valkey-unit-tests
-$(ENGINE_UNIT_TESTS): $(ENGINE_TEST_OBJ) $(ENGINE_LIB_NAME)
+$(ENGINE_UNIT_TESTS): $(ENGINE_TEST_OBJ) $(ENGINE_LIB_NAME) valkey-benchmark-dataset.o
```
similar to the cmake file, I feel like this should be part of a obj list or something, right? Is there a good reason it's not?

VoletiRam replied:
For unit testing, we only use APis in valkey-benchmark-dataset and we don't need benchmark binary (or rest of the objects). This is the minimal addition to add unit testing for dataset.

rainsupreme commented on src/valkey-benchmark-dataset.h:
```diff
@@ -0,0 +1,86 @@
+/* Dataset support for valkey-benchmark
+ *
+ * Copyright (c) 2009-2012, Redis Ltd.
```
we are not redis

VoletiRam replied:
Ack.

rainsupreme commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,776 @@
+/* Dataset support for valkey-benchmark
+ *
+ * Copyright (c) 2009-2012, Redis Ltd.
```
we are not redis

VoletiRam replied:
Ack.

rainsupreme commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,776 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
This license/copyright boilerplate is unnecessary - you can use a short header like the one in `hashtable.c`. This applies to all the other files as well.

```
/*
 * Copyright Valkey Contributors.
 * All rights reserved.
 * SPDX-License-Identifier: BSD 3-Clause
 */
```

rainsupreme commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,776 @@
+}
+
+static sds getXmlFieldValue(const char *xml_doc, const char *field_name) {
+    char start_tag_prefix[128], end_tag[128];
```
So there's limit on field length. Is there input checking for this and a uni test? If this length comes up other places then it should be a #define constant.

VoletiRam replied:
Will increase this limt as well.

rainsupreme commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,776 @@
+    return sdsnewlen(content_start, content_len);
+}
+
+static int csvDiscoverFields(dataset *ds) {
```
you can use `bool` return type if you `#include <stdbool.h>` - this helps to clarify what the return value means. (and then of course use `true` and `false` instead of 1 and 0). Best to update the whole file with bool for consistency.

VoletiRam replied:
Ack. Will update.

rainsupreme commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,776 @@
+}
+
+static int scanXmlFields(const char *doc_start, const char *doc_end, dataset *ds, const char *start_root_tag, const char *end_root_tag) {
+    char field_names[MAX_DATASET_FIELDS][64];
```
now there's a 64 character size limit here? Feels like magic numbers. I'd prefer to see a central #define for these size limits.

I see other 64 character limits below too

VoletiRam replied:
I want to increase this limit just to be future proof for large field testing.

rainsupreme reviewed (changes requested):
I had a couple questions about the makefile stuff, but mainly the copyright/license boilerplate headers need to be fixed.

The refactor to a separate file looks good to me though. :)

Commit:
Address the comments on cmake file, field size and count limitation

Address the comments on cmake file, field size and count limitations.
Update return values to use bool. Increase limits to be future proof.

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Merge remote-tracking branch 'origin/unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme commented on src/Makefile:
```diff
@@ -586,6 +586,8 @@ ENGINE_CLI_OBJ = \
     zmalloc.o
 ENGINE_BENCHMARK_NAME=$(ENGINE_NAME)-benchmark$(PROG_SUFFIX)
+# Dataset module shared between benchmark and unit tests
+BENCHMARK_DATASET_OBJ=valkey-benchmark-dataset.o
```
I guess I had imagined adding it to ENGINE_BENCHMARK_OBJ, but I see now that it is used in UTs too. I'm not sure a list with a single item makes sense - perhaps we should just add it to both ENGINE_BENCHMARK_OBJ and ENGINE_UNIT_TESTS instead? 🤔

VoletiRam replied:
Like before?

rainsupreme replied:
Not quite what I meant - before in cmake we had this:
`add_executable(valkey-unit-tests ${UNIT_TEST_SRCS} ${CMAKE_SOURCE_DIR}/src/valkey-benchmark-dataset.c)`
My suggestion is to also add it to the UNIT_TEST_SRCS list.

For the makefile, this is the previous line:
`$(ENGINE_UNIT_TESTS): $(ENGINE_TEST_OBJ) $(ENGINE_LIB_NAME) valkey-benchmark-dataset.o`
Could we put it in ENGINE_TEST_OBJ?

Let me know if there's a reason that's impractical or if you disagree, but that would seem better organized to me. Also, feel free to get a second opinion 🙂

VoletiRam replied:
Took some time to understand the suggestion. Makes sense. I update the PR and along with cluster mode enhancement.

rainsupreme reviewed (changes requested):
The new changes look good to me. Just one or two quick fixes away now! :)

Commit:
Address comments on license and make file

Address comments on license and make file

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme reviewed (approved):
I made a suggestion for the make files but it's kind of a nit - I have no objection to merging it now. :)

Commit:
Fix dataset cmake and add tag support for cluster mode

Fix dataset cmake and add tag support for cluster mode

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme commented on src/valkey-benchmark.c:
```diff
@@ -583,6 +583,24 @@ static void resetClient(client c) {
 }
 
+/* Scan buffer for {tag} placeholders and store positions */
+static void scanClusterTags(client c, char *buffer_start) {
```
Cover tag support in a test?

VoletiRam replied:
Just added.

Commit:
Add cluster tag test for dataset feature in benchmark

Add cluster tag test for dataset feature in benchmakr

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix test for ubuntu-latest

Fix test for ubuntu-latest

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

madolson commented on src/unit/test_dataset.cpp:
```diff
@@ -0,0 +1,201 @@
+#include <stdio.h>
```
Do we need this unit test? This feels like a file that will primarily benefit from integration testing.

VoletiRam replied:
Unit tests work well here because they test parsing logic in isolation without server overhead. They're fast and great for edge cases like validating 513-character field name rejection or the 1000/1001 field boundary—scenarios that are hard to test reliably through integration tests.

madolson replied:
We can leave them. I think I was originally concerned about all the custom parsing, less about the tests.

VoletiRam replied:
Some of the limits enforced before are only on XML parsing which lead to writing unit tests as it is not possible to test them with integration tests.

 Non-XML related unit tests were `field_mapping_optimization` and `field_count_correctness` are covered implicitly by all existing dataset integration tests that verify field values are correctly inserted. Felt like good to have's so removed them.

`csv_quoted_fields` -  this was the one genuine edge case not covered by integration before. It has been now added to `tests/integration/valkey-benchmark.tcl` as `benchmark: dataset CSV with quoted field values containing commas`, which verifies the value `"Book, Part 1"` is stored intact end-to-end.

The unit test file also introduced a non-trivial build dependency: `valkey-benchmark-dataset.o` had to be compiled separately and linked into the gtest binary, adding coupling between the benchmark build and the server unit test infrastructure. Removing it simplifies both `src/unit/Makefile` and `src/unit/CMakeLists.txt`

madolson commented on src/Makefile:
```diff
@@ -31,7 +31,7 @@ endif
 	OPTIMIZATION+=-fno-omit-frame-pointer
 endif
-DEPENDENCY_TARGETS=libvalkey linenoise hdr_histogram fpconv
+DEPENDENCY_TARGETS=libvalkey linenoise lua hdr_histogram fpconv
```
I don't understand these changes. Why are they needed?

VoletiRam replied:
this was left from merge commit. Will remove it.

Commit:
Reuse stream buffer logic in loading too for xml and address comments of PR

Reuse stream buffer logic in loading too for xml and address comments of
PR

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Merge remote-tracking branch 'origin/unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

VoletiRam commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,782 @@
+        current_doc = sdscat(current_doc, buffer);
+
+        const char *tag_prefix_pos = strstr(current_doc, start_tag_prefix);
+        if (!tag_prefix_pos) continue;
```
I don't want to limit the doc growth as I expect the code to work with larger datasets. Will leave the handling of malformed datasets and potential system constraints to users. 


Neverthless, this comment helped me identify potential reuse of stream scan logic in scan too which I addressed.

Commit:
Fix unit test

Fix unit test

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

VoletiRam commented:
Failure appears to be unrelated. 
```

*** [TIMEOUT]: replica do not write the reply to the replication link - PSYNC (_addReplyToBufferOrList) in tests/integration/replication.tcl
Cleanup: may take some time... OK
make[1]: *** [Makefile:808: lcov] Error 1
make[1]: Leaving directory '/home/runner/work/valkey/valkey/src'
make: *** [Makefile:6: lcov] Error 2
Error: Process completed with exit code 2.
```

madolson commented on src/valkey-benchmark-dataset.c:
```diff
@@ -0,0 +1,843 @@
+/* Dataset support for valkey-benchmark
```
Do we need custom XML, CSV, and TSV parsing here? It seems trivially easy to go and just reformat the code to one of those three?

VoletiRam replied:
We need XML to support comprehensive data formats like vectors which can be stored with encodings such as base64/hex. 

While users could convert their data, I am aimed this feature to be a turnkey solution. By supporting the native formats of the most popular search (XML/Wikipedia dumps for text) and vector (CSV/TSV) datasets, we eliminate the pre-processing/converting bottleneck for majority of datasets available out there. We included CSV/TSV as the default for simple, tabular SET/GET workloads because it is the most straightforward mapping for command arguments.  Together, these three cover 99% of existing public datasets used for database performance testing. The idea is to make the tool as frictionless as possible where users can link datasets straight out downloading while support simple cases with csv without much overhead on parsing.

madolson replied:
> While users could convert their data, I am aimed this feature to be a turnkey solution. By supporting the native formats of the most popular search (XML/Wikipedia dumps for text) and vector (CSV/TSV) datasets, we eliminate the pre-processing/converting bottleneck for majority of datasets available out there. We included CSV/TSV as the default for simple, tabular SET/GET workloads because it is the most straightforward mapping for command arguments. Together, these three cover 99% of existing public datasets used for database performance testing. The idea is to make the tool as frictionless as possible where users can link datasets straight out downloading while support simple cases with csv without much overhead on parsing.

This seems like a tiny python code to translate it. Let's pick a format that makes the most sense and just use that. I would rather not maintain a bunch of code.

VoletiRam replied:
Cool. I removed all XML related logic and picked CSV/TSV as default format. Please take a look when you have time. Thank you.

madolson commented on src/valkey-benchmark.c:
```diff
@@ -2220,12 +2338,59 @@ int main(int argc, char **argv) {
+
+        if (verbose) {
+            datasetReportMemory(config.current_dataset);
+        }
```
I'm not really sure about the point of this API. Why do we need to print out the dataset memory usage?

VoletiRam replied:
I initially added this to validate whether we load all the data as expected. Will remove it.

Commit:
Merge remote-tracking branch 'origin/unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix github wf issues with build and clang

Fix github wf issues with build and clang

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

rainsupreme reviewed (approved):
LGTM! I realize that this feature can be used to test GET/SET and other commands too, not just full text search. I look forward to the documentation being updated on the Valkey website 😉

madolson reviewed (commented):
Meta comment is do we really need all of the complex parsing? Can we just use XML and then do conversion in a higher level language like python?

Commit:
Address comments on PR to remove memory report

Address comments on PR to remove memory report, fix makefile lua
addition, and make explicit building of dataset in unit test instead

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Fix codecov

Fix codecov

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Remove XML support and related code

Remove XML support and related code

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

Commit:
Merge remote-tracking branch 'origin/unstable' into dataset-support

Signed-off-by: Ram Prasad Voleti <ramvolet@amazon.com>

VoletiRam commented:
All the test failures seem unrelated

madolson commented:
There are some test failures, so going to wait to merge until after they're resolved for the moment. @VoletiRam can you update the top level comment?

VoletiRam committed:
Merge branch 'unstable' into dataset-support

VoletiRam committed:
Merge branch 'unstable' into dataset-support
